File: backend/app/api/votes.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.democracy import VoteAndInitiative

router = APIRouter(prefix="/votes", tags=["democracy"])


class VoteOut:
    pass  # inline below


from pydantic import BaseModel
from datetime import date


class VoteOut(BaseModel):
    id: uuid.UUID
    vote_type: str
    level: str
    canton: str | None = None
    title_de: str | None = None
    title_fr: str | None = None
    title_it: str | None = None
    official_url: str | None = None
    vote_date: date | None = None
    status: str
    result: dict | None = None
    pro_arguments: dict | None = None
    contra_arguments: dict | None = None

    model_config = {"from_attributes": True}


@router.get("", response_model=list[VoteOut])
async def list_votes(
    status: str | None = None,
    level: str | None = None,
    canton: str | None = None,
    limit: int = Query(default=20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
):
    query = select(VoteAndInitiative)
    if status:
        query = query.where(VoteAndInitiative.status == status)
    if level:
        query = query.where(VoteAndInitiative.level == level)
    if canton:
        query = query.where(VoteAndInitiative.canton == canton)
    query = query.order_by(VoteAndInitiative.vote_date.desc()).limit(limit)
    result = await db.execute(query)
    return [VoteOut.model_validate(v) for v in result.scalars().all()]


@router.get("/{vote_id}", response_model=VoteOut)
async def get_vote(vote_id: uuid.UUID, db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(VoteAndInitiative).where(VoteAndInitiative.id == vote_id)
    )
    vote = result.scalar_one_or_none()
    if not vote:
        raise HTTPException(status_code=404, detail="Vote not found")
    return VoteOut.model_validate(vote)

File: worker/src/routes/votes.ts
import { Hono } from 'hono';
import type { Env } from '../types';
import { parseJson } from '../utils';

export function votesRoutes() {
  const app = new Hono<{ Bindings: Env }>();

  app.get('/', async (c) => {
    const status = c.req.query('status');
    const level = c.req.query('level');
    const canton = c.req.query('canton');
    const limit = Math.min(parseInt(c.req.query('limit') || '20'), 100);

    let where = 'WHERE 1=1';
    const params: unknown[] = [];
    if (status) { where += ' AND status = ?'; params.push(status); }
    if (level) { where += ' AND level = ?'; params.push(level); }
    if (canton) { where += ' AND canton = ?'; params.push(canton); }

    const rows = await c.env.DB.prepare(
      `SELECT * FROM votes_and_initiatives ${where} ORDER BY vote_date DESC LIMIT ?`
    ).bind(...params, limit).all();

    return c.json(rows.results.map((v) => ({
      id: v.id, vote_type: v.vote_type, level: v.level, canton: v.canton,
      title_de: v.title_de, title_fr: v.title_fr, title_it: v.title_it,
      official_url: v.official_url, vote_date: v.vote_date, status: v.status,
      result: parseJson(v.result as string, null),
      pro_arguments: parseJson(v.pro_arguments as string, null),
      contra_arguments: parseJson(v.contra_arguments as string, null),
    })));
  });

  app.get('/:id', async (c) => {
    const row = await c.env.DB.prepare('SELECT * FROM votes_and_initiatives WHERE id = ?').bind(c.req.param('id')).first();
    if (!row) return c.json({ error: 'Not found' }, 404);
    return c.json({
      id: row.id, vote_type: row.vote_type, level: row.level, canton: row.canton,
      title_de: row.title_de, title_fr: row.title_fr, title_it: row.title_it,
      official_url: row.official_url, vote_date: row.vote_date, status: row.status,
      result: parseJson(row.result as string, null),
      pro_arguments: parseJson(row.pro_arguments as string, null),
      contra_arguments: parseJson(row.contra_arguments as string, null),
      financial_impact: parseJson(row.financial_impact as string, null),
    });
  });

  return app;
}
